source/_posts/2016-10-18-Understanding-code-signing-for-iOS-apps.md

```diff
@@ -0,0 +1,91 @@
+date: 2016-10-18 23:10:57
+tags: ios,code,signing,security,certificates,apns
+authorId: maco
+categories:
+- iOS
+---
+
+If you're an iOS developer, most likely you've had to deal with code signing. And if you're a junior iOS developer, you might've felt a bit overwhelmed by everything going on in the "Certificates, Identifiers & Profiles" section of the developer portal.
+
+<p align=center><img src="https://d1gwekl0pol55k.cloudfront.net/image/baas/translate_values/hbikk_ospmSpNLyW.gif" /></p>
+
+The goal of this article is to help junior iOS developers understand code signing on a higher level. This will not be a step-by-step tutorial on how to code sign your app. Ideally, after reading this article, you would be able to code sign your app without following any tutorials.
+
+I don't plan to go into lower level details, but we will talk a bit about asymmetric cryptography. 
+
+### Asymmetric cryptography
+The minimum you need to know is that asymmetric cryptography uses a **public key** and a **private key**. The users have to keep their private key for themselves, but they can share the public key. And using those public and private keys, a user can prove that he is indeed himself.
+
+A good high-level explanation of asymmetric cryptography can be found [here](https://blog.vrypan.net/2013/08/28/public-key-cryptography-for-non-geeks/). If you want to know implementation details or the math behind this, you can find them online.
+
+
+### App id
+The app id is the unique identifier of your app. It consists of a team id, generated by Apple (you don't have any control over it) and your app's bundle id (`com.youcompany.yourapp`, for example). 
+
+There can also be wildcard app ids: `com.yourcompany.*`. Those will match on multiple bundle ids. 
+
+Generally, you will have an explicit app id, not a wildcard one.
+
+### Certificates
+You have probably already noticed that in order to create a certificate in Apple's developer portal, you need to upload a Certificate Signing Request. You can generate this CSR from the Keychain, and this CSR contains a private key.
+
+Then on the developer portal, you can create a certificate using this CSR. 
+
+There can be multiple types of certificates. The most common are:
+
+* Development (iOS App Development) - You need those to run your app on a device from Xcode. 
+* Distribution (App Store and Ad Hoc) - You need those to be able to distribute your app through the App Store or Ad Hoc
+* APNS (Apple Push Notification Service)  - You need those to be able to send push notifications to your app. Unlike Development or Distribution certificates, APNS certificates are associated with an app id. There can be 2 types of APNS certificates, for development - Apple Push Notification service SSL (Sandbox), and for production - Apple Push Notification service SSL (Sandbox & Production). You need to create both of them if you want push notifications to work on both debug and distribution builds. 
```

Comment:
Yes that's the one. Sorry I thought you were talking about that one :D Think it would be nice to mention since it makes the whole process hugely easier.